<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from model.urdf.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="simple_car">
  <link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1580"/>
      <inertia ixx="263.3333333333333" ixy="0" ixz="0" iyy="263.3333333333333" iyz="0" izz="263.3333333333333"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="1 1 1"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="1 1 1"/>
      </geometry>
    </collision>
  </link>
  <link name="wheel_fl_base">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_fl_steer">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_fl_tire">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="200"/>
      <inertia ixx="15.417916666666665" ixy="0" ixz="0" iyy="15.417916666666665" iyz="0" izz="25.5025"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!-- 
                    <cylinder radius="${radius}" length="${length}"/>
-->
        <mesh filename="model://simple_car/resources/tire/tire.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.4" radius="0.505"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_base_link_to_wheel_fl_base" type="fixed">
    <origin rpy="0 0 0" xyz="0.2 1.2 0"/>
    <parent link="base_link"/>
    <child link="wheel_fl_base"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000000000000000" lower="-1" upper="1" velocity="100"/>
    <safety_controller k_position="1" k_velocity="1" soft_lower_limit="0" soft_upper_limit="0"/>
  </joint>
  <joint name="joint_wheel_fl_base_to_wheel_fl_steer" type="continuous">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="wheel_fl_base"/>
    <child link="wheel_fl_steer"/>
    <axis xyz="0 0 1"/>
  </joint>
  <joint name="joint_wheel_fl_steer_to_wheel_fl_tire" type="continuous">
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0 -0.655"/>
    <parent link="wheel_fl_steer"/>
    <child link="wheel_fl_tire"/>
    <axis xyz="0 0 1"/>
  </joint>
  <!--
            <transmission name="${name}_steer_transmission">
                <type>transmission_interface/SimpleTransmission</type>
                <actuator name="$${name}_steer_motor">
                    <mechanicalReduction>1</mechanicalReduction>
                </actuator>
                <joint name="joint_${name}_base_to_${name}_steer">
                    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
                </joint>
            </transmission>
        -->
  <gazebo reference="wheel_fl_steer">
    <material>Gazebo/Blue</material>
  </gazebo>
  <!--
        <gazebo reference="${name}_tire">
            <material>Gazebo/Orange</material>
            <mu1>0.9</mu1>
            <mu2>0.8</mu2>
        </gazebo>
        -->
  <gazebo>
    <plugin filename="libwheel.so" name="wheel_fl">
      <steer_joint name="joint_wheel_fl_base_to_wheel_fl_steer"/>
      <drive_joint name="joint_wheel_fl_steer_to_wheel_fl_tire"/>
    </plugin>
  </gazebo>
  <link name="wheel_rl_base">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_rl_steer">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_rl_tire">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="200"/>
      <inertia ixx="15.417916666666665" ixy="0" ixz="0" iyy="15.417916666666665" iyz="0" izz="25.5025"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!-- 
                    <cylinder radius="${radius}" length="${length}"/>
-->
        <mesh filename="model://simple_car/resources/tire/tire.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.4" radius="0.505"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_base_link_to_wheel_rl_base" type="fixed">
    <origin rpy="0 0 0" xyz="-3 1.2 0"/>
    <parent link="base_link"/>
    <child link="wheel_rl_base"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000000000000000" lower="-1" upper="1" velocity="100"/>
    <safety_controller k_position="1" k_velocity="1" soft_lower_limit="0" soft_upper_limit="0"/>
  </joint>
  <joint name="joint_wheel_rl_base_to_wheel_rl_steer" type="continuous">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="wheel_rl_base"/>
    <child link="wheel_rl_steer"/>
    <axis xyz="0 0 1"/>
  </joint>
  <joint name="joint_wheel_rl_steer_to_wheel_rl_tire" type="continuous">
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0 -0.655"/>
    <parent link="wheel_rl_steer"/>
    <child link="wheel_rl_tire"/>
    <axis xyz="0 0 1"/>
  </joint>
  <!--
            <transmission name="${name}_steer_transmission">
                <type>transmission_interface/SimpleTransmission</type>
                <actuator name="$${name}_steer_motor">
                    <mechanicalReduction>1</mechanicalReduction>
                </actuator>
                <joint name="joint_${name}_base_to_${name}_steer">
                    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
                </joint>
            </transmission>
        -->
  <gazebo reference="wheel_rl_steer">
    <material>Gazebo/Blue</material>
  </gazebo>
  <!--
        <gazebo reference="${name}_tire">
            <material>Gazebo/Orange</material>
            <mu1>0.9</mu1>
            <mu2>0.8</mu2>
        </gazebo>
        -->
  <gazebo>
    <plugin filename="libwheel.so" name="wheel_rl">
      <steer_joint name="joint_wheel_rl_base_to_wheel_rl_steer"/>
      <drive_joint name="joint_wheel_rl_steer_to_wheel_rl_tire"/>
    </plugin>
  </gazebo>
  <link name="wheel_rr_base">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_rr_steer">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_rr_tire">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="200"/>
      <inertia ixx="15.417916666666665" ixy="0" ixz="0" iyy="15.417916666666665" iyz="0" izz="25.5025"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!-- 
                    <cylinder radius="${radius}" length="${length}"/>
-->
        <mesh filename="model://simple_car/resources/tire/tire.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.4" radius="0.505"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_base_link_to_wheel_rr_base" type="fixed">
    <origin rpy="0 0 0" xyz="-3 -1.2 0"/>
    <parent link="base_link"/>
    <child link="wheel_rr_base"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000000000000000" lower="-1" upper="1" velocity="100"/>
    <safety_controller k_position="1" k_velocity="1" soft_lower_limit="0" soft_upper_limit="0"/>
  </joint>
  <joint name="joint_wheel_rr_base_to_wheel_rr_steer" type="continuous">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="wheel_rr_base"/>
    <child link="wheel_rr_steer"/>
    <axis xyz="0 0 1"/>
  </joint>
  <joint name="joint_wheel_rr_steer_to_wheel_rr_tire" type="continuous">
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0 -0.655"/>
    <parent link="wheel_rr_steer"/>
    <child link="wheel_rr_tire"/>
    <axis xyz="0 0 1"/>
  </joint>
  <!--
            <transmission name="${name}_steer_transmission">
                <type>transmission_interface/SimpleTransmission</type>
                <actuator name="$${name}_steer_motor">
                    <mechanicalReduction>1</mechanicalReduction>
                </actuator>
                <joint name="joint_${name}_base_to_${name}_steer">
                    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
                </joint>
            </transmission>
        -->
  <gazebo reference="wheel_rr_steer">
    <material>Gazebo/Blue</material>
  </gazebo>
  <!--
        <gazebo reference="${name}_tire">
            <material>Gazebo/Orange</material>
            <mu1>0.9</mu1>
            <mu2>0.8</mu2>
        </gazebo>
        -->
  <gazebo>
    <plugin filename="libwheel.so" name="wheel_rr">
      <steer_joint name="joint_wheel_rr_base_to_wheel_rr_steer"/>
      <drive_joint name="joint_wheel_rr_steer_to_wheel_rr_tire"/>
    </plugin>
  </gazebo>
  <link name="wheel_fr_base">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_fr_steer">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.075"/>
      <geometry>
        <box size="0.3 0.3 0.15"/>
      </geometry>
    </visual>
  </link>
  <link name="wheel_fr_tire">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="200"/>
      <inertia ixx="15.417916666666665" ixy="0" ixz="0" iyy="15.417916666666665" iyz="0" izz="25.5025"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!-- 
                    <cylinder radius="${radius}" length="${length}"/>
-->
        <mesh filename="model://simple_car/resources/tire/tire.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.4" radius="0.505"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_base_link_to_wheel_fr_base" type="fixed">
    <origin rpy="0 0 0" xyz="0.2 -1.2 0"/>
    <parent link="base_link"/>
    <child link="wheel_fr_base"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000000000000000" lower="-1" upper="1" velocity="100"/>
    <safety_controller k_position="1" k_velocity="1" soft_lower_limit="0" soft_upper_limit="0"/>
  </joint>
  <joint name="joint_wheel_fr_base_to_wheel_fr_steer" type="continuous">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="wheel_fr_base"/>
    <child link="wheel_fr_steer"/>
    <axis xyz="0 0 1"/>
  </joint>
  <joint name="joint_wheel_fr_steer_to_wheel_fr_tire" type="continuous">
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0 -0.655"/>
    <parent link="wheel_fr_steer"/>
    <child link="wheel_fr_tire"/>
    <axis xyz="0 0 1"/>
  </joint>
  <!--
            <transmission name="${name}_steer_transmission">
                <type>transmission_interface/SimpleTransmission</type>
                <actuator name="$${name}_steer_motor">
                    <mechanicalReduction>1</mechanicalReduction>
                </actuator>
                <joint name="joint_${name}_base_to_${name}_steer">
                    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
                </joint>
            </transmission>
        -->
  <gazebo reference="wheel_fr_steer">
    <material>Gazebo/Blue</material>
  </gazebo>
  <!--
        <gazebo reference="${name}_tire">
            <material>Gazebo/Orange</material>
            <mu1>0.9</mu1>
            <mu2>0.8</mu2>
        </gazebo>
        -->
  <gazebo>
    <plugin filename="libwheel.so" name="wheel_fr">
      <steer_joint name="joint_wheel_fr_base_to_wheel_fr_steer"/>
      <drive_joint name="joint_wheel_fr_steer_to_wheel_fr_tire"/>
    </plugin>
  </gazebo>
  <link name="tool_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="400"/>
      <inertia ixx="722.9999999999999" ixy="0" ixz="0" iyy="56.33333333333332" iyz="0" izz="683.3333333333333"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="model://simple_car/resources/tire/tire.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.5 4.5 1.2"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_tool_to_base_base" type="fixed">
    <origin rpy="0 0 0" xyz="20 0 0"/>
    <parent link="base_link"/>
    <child link="tool_link"/>
    <axis xyz="0 0 1"/>
  </joint>
  <gazebo>
    <plugin filename="libvehicle_body.so" name="vehicle">
        </plugin>
  </gazebo>
</robot>
